feat: set up automated smoke tests for supported Operating Systems

kruskall opened the pull request:
<!-- Thanks for sending a pull request!

If this is your first contribution, please review and sign our contributor agreement -
https://www.elastic.co/contributor-agreement.

Guidelines:
 - Prefer small PRs, and split changes into multiple logical commits where they must
   be delivered in a single PR.
 - If the PR is incomplete and not yet ready for review, open it as a Draft.
 - Once the PR is marked ready for review it is expected to pass all tests and linting,
   and you should not force-push any changes.

See also https://github.com/elastic/apm-server/blob/main/CONTRIBUTING.md for more tips on contributing.
-->

## Motivation/summary

Add smoke testing for the following os (from the support matrix):

- ubuntu 18.04
- ubuntu 20.04
- ubuntu 22.04
- debian 10
- debian 11
- Amazon Linux 2
- RHEL 7
- RHEL 8

The terraform script provisions an ec2 instance and downloads the
apm server from the artifacts api.
The apm-server is installed and linked to the es cluster.

## Checklist

<!--
Delete irrelevant items. The changelog should only be updated for user-facing changes.
Once the PR is ready for review there should be no unticked boxes.
-->

- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)
- [ ] Documentation has been updated

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes

<!--
Explain how this PR can be tested by the reviewer: commands, dependencies, steps, etc.
If it is self-explanatory, delete this section.
-->

## Related issues

Related to https://github.com/elastic/apm-server/issues/9184

apmmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-9440/detail/PR-9440/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-9440/detail/PR-9440/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-9440/detail/PR-9440/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-server%2Fapm-server-mbp%2FPR-9440/detail/PR-9440/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-server_9440.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/apm-ci/transactions/view?rangeFrom=2022-10-26T04:13:41.805Z&rangeTo=2022-10-26T04:33:41.805Z&transactionName=BUILD+apm-server%2Fapm-server-mbp%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=5f34b6fbb4853965ea0609f61ac4e33a&transactionId=8b2571b043ae8581)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-10-26T04:23:41.805+0000


* Duration: 27 min 5 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 154  |
| Skipped      | 0 |
| Total        | 154   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate and publish the docker images.


- `/test windows` : Build & tests on Windows.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

apmmachine commented:
### :books: Go benchmark report

Diff with the `main` branch

```
name                                                                                              old time/op    new time/op    delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/beater/request goos:linux goarch:amd64
ContextResetContentEncoding/empty-12                                                                 118ns ± 1%     121ns ± 2%   +2.05%  (p=0.008 n=5+5)
ContextResetContentEncoding/uncompressed-12                                                          152ns ± 1%     142ns ± 2%   -6.93%  (p=0.008 n=5+5)
ContextResetContentEncoding/deflate-12                                                              3.41µs ±52%    4.11µs ± 2%  +20.63%  (p=0.016 n=5+5)
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
BackendProcessor/invalid-event.ndjson-12                                                            32.2µs ±13%    26.1µs ±15%  -19.11%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2/spans.ndjson-12                         34.3µs ±13%    39.2µs ±14%  +14.11%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel4/minimal.ndjson-12                       4.24µs ± 5%    4.49µs ± 3%   +5.77%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel4/ratelimit.ndjson-12                     11.0µs ± 2%    11.9µs ± 8%   +7.58%  (p=0.016 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8/errors_2.ndjson-12                      8.09µs ± 2%    7.94µs ± 1%   -1.81%  (p=0.016 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8/metadata-null-values.ndjson-12          1.17µs ± 2%    1.15µs ± 1%   -1.81%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8/metricsets.ndjson-12                    4.96µs ± 1%    5.04µs ± 1%   +1.65%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/invalid-metadata-2.ndjson-12           468ns ± 1%     489ns ± 1%   +4.41%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/invalid-metadata.ndjson-12             476ns ± 1%     493ns ± 1%   +3.56%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/metadata-null-values.ndjson-12         762ns ± 1%     774ns ± 1%   +1.49%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/optional-timestamps.ndjson-12         1.45µs ± 2%    1.42µs ± 1%   -1.90%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/transactions_spans.ndjson-12          9.37µs ± 2%    9.16µs ± 2%   -2.29%  (p=0.032 n=5+5)
ReadBatch/metadata.ndjson-12                                                                        13.0µs ±14%    19.0µs ±14%  +45.73%  (p=0.008 n=5+5)
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
TraceGroups-12                                                                                       142ns ± 2%     122ns ± 2%  -14.27%  (p=0.016 n=5+4)
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64
ReadEvents/json_codec/199_events-12                                                                 1.57ms ± 5%    1.48ms ± 2%   -6.13%  (p=0.016 n=5+5)
ReadEvents/json_codec/399_events-12                                                                 3.07ms ± 3%    2.85ms ± 1%   -6.95%  (p=0.008 n=5+5)
ReadEvents/json_codec/1000_events-12                                                                7.42ms ± 5%    7.17ms ± 3%   -3.40%  (p=0.032 n=5+5)

name                                                                                              old alloc/op   new alloc/op   delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/beater/request goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
BackendProcessor/transactions_spans.ndjson-12                                                       24.1kB ± 1%    24.3kB ± 0%   +1.20%  (p=0.016 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2/errors_transaction_id.ndjson-12         19.9kB ± 1%    20.2kB ± 0%   +1.34%  (p=0.016 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2/metadata-null-values.ndjson-12          3.69kB ± 1%    3.74kB ± 2%   +1.55%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2/metadata.ndjson-12                      5.32kB ± 1%    5.39kB ± 1%   +1.25%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel4/invalid-json-metadata.ndjson-12         5.88kB ± 1%    5.97kB ± 2%   +1.55%  (p=0.016 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8/invalid-event.ndjson-12                 8.36kB ± 2%    8.25kB ± 1%   -1.29%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8/metadata-null-values.ndjson-12          3.98kB ± 1%    3.91kB ± 1%   -1.69%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/invalid-metadata-2.ndjson-12          3.13kB ± 1%    3.09kB ± 1%   -1.46%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/otel-bridge.ndjson-12                 11.4kB ± 1%    11.5kB ± 0%   +0.79%  (p=0.032 n=5+5)
ReadBatch/heavy.ndjson-12                                                                            839kB ± 0%     838kB ± 0%   -0.10%  (p=0.032 n=5+5)
ReadBatch/transactions_spans.ndjson-12                                                              22.8kB ± 0%    22.8kB ± 0%   -0.07%  (p=0.008 n=5+5)
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64
ReadEvents/json_codec/199_events-12                                                                 1.10MB ± 0%    1.10MB ± 0%   -0.07%  (p=0.040 n=5+5)

name                                                                                              old allocs/op  new allocs/op  delta
pkg:github.com/elastic/apm-server/internal/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/beater/request goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
ReadBatch/heavy.ndjson-12                                                                            22.1k ± 0%     22.1k ± 0%   -0.01%  (p=0.000 n=5+4)
pkg:github.com/elastic/apm-server/internal/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64

name                                                                                              old speed      new speed      delta
pkg:github.com/elastic/apm-server/internal/processor/stream goos:linux goarch:amd64
BackendProcessor/invalid-event.ndjson-12                                                          24.0MB/s ±15%  29.6MB/s ±16%  +23.56%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2/events.ndjson-12                       170MB/s ± 4%   147MB/s ±16%  -13.66%  (p=0.048 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2/spans.ndjson-12                        236MB/s ±12%   207MB/s ±15%  -12.27%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel4/minimal.ndjson-12                      243MB/s ± 5%   230MB/s ± 3%   -5.55%  (p=0.016 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel4/ratelimit.ndjson-12                    382MB/s ± 2%   356MB/s ± 8%   -6.89%  (p=0.016 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8/errors_2.ndjson-12                     583MB/s ± 2%   594MB/s ± 1%   +1.84%  (p=0.016 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8/metadata-null-values.ndjson-12         450MB/s ± 2%   459MB/s ± 1%   +1.86%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8/metricsets.ndjson-12                   513MB/s ± 1%   505MB/s ± 1%   -1.62%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/invalid-metadata-2.ndjson-12         932MB/s ± 1%   892MB/s ± 1%   -4.23%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/invalid-metadata.ndjson-12           937MB/s ± 1%   905MB/s ± 1%   -3.44%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/metadata-null-values.ndjson-12       690MB/s ± 1%   680MB/s ± 1%   -1.48%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/optional-timestamps.ndjson-12        707MB/s ± 2%   721MB/s ± 1%   +1.93%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel200/transactions_spans.ndjson-12         621MB/s ± 2%   635MB/s ± 2%   +2.32%  (p=0.032 n=5+5)
ReadBatch/metadata.ndjson-12                                                                      96.4MB/s ±16%  66.1MB/s ±15%  -31.39%  (p=0.008 n=5+5)

```

_report generated with https://pkg.go.dev/golang.org/x/perf/cmd/benchstat_
<!--COMMENT_GENERATED_WITH_ID_bench.diff-->

axw commented on testing/smoke/supported-os/test.sh:
```diff
@@ -0,0 +1,65 @@
+
+OBSERVER_VERSION=$(echo ${VERSION} | cut -d '-' -f1 )
+
+if [[ ${MAJOR_VERSION} -eq 7 ]]; then
```
Are we testing old versions of APM Server or the stack? Is this code path relevant?

kruskall replied:
I don't think we are. Good catch! :+1:

axw commented on testing/smoke/supported-os/test.sh:
```diff
@@ -0,0 +1,65 @@
+
+for os in "${os_names[@]}"
+do
+    terraform_apply ${LATEST_VERSION} ${INTEGRATIONS_SERVER} ${KEY_NAME} $os
+    healthcheck 1
+    send_events
+    ${ASSERT_EVENTS_FUNC} ${OBSERVER_VERSION}
+    terraform_destroy
```
Applying and destroying in a loop means we'll create the whole ESS deployment from fresh each time, right? That sounds like it'll be a bit slow.

Would it work to apply each time, delete data streams at the end of each OS's tests, and destroy right at the end? Subsequent applies should update changed and delete unused resources, right?

kruskall replied:
Good point! I've update the code to reuse the cluster :+1:

axw reviewed (commented):
Looks nice overall. I think mixing the ESS and OS smoke test Terraform together might be a recipe for complexity, though.

None of the AWS instance stuff is relevant to the ESS smoke tests. Some of the ESS things (secret token, server URL, default stack version) are not relevant to the OS smoke tests. So, would we be better of creating a dedicated Terraform module for the OS smoke tests? That way we can avoid conditionals, and minimise the amount of context required to understand them.

If we had a Terraform module just for the OS smoke tests we would create an ESS deployment with Elasticsearch, Kibana, and Integrations Server for its Fleet Server component. The ESS deployment would always be 8.x deployment, as we won't be doing automated OS smoke tests for 7.x.

kruskall commented:
> Looks nice overall. I think mixing the ESS and OS smoke test Terraform together might be a recipe for complexity, though.
> 
> None of the AWS instance stuff is relevant to the ESS smoke tests. Some of the ESS things (secret token, server URL, default stack version) are not relevant to the OS smoke tests. So, would we be better of creating a dedicated Terraform module for the OS smoke tests? That way we can avoid conditionals, and minimise the amount of context required to understand them.

Does it need a module ? All the unrelated stuff are in the `testing/smoke/main.tf` script. I think we can achieve the same result by using a separate tf script without the conditionals.

axw commented:
> Does it need a module ? All the unrelated stuff are in the testing/smoke/main.tf script. I think we can achieve the same result by using a separate tf script without the conditionals.

@kruskall I _think_ we're saying the same thing. What I meant was to create a separate `.tf` file, and not reuse `testing/smoke/main.tf` - since that one is more appropriate for testing APM Server running in ESS.

kruskall commented:
> @kruskall I _think_ we're saying the same thing. What I meant was to create a separate `.tf` file, and not reuse `testing/smoke/main.tf` - since that one is more appropriate for testing APM Server running in ESS.

Yep, that's right, I thought you wanted to create a new `module`, similar to `ec_deployment`.
I've updated to code to use a separate file, removed the `count` field with the indexes and the conditionals.

axw commented on testing/smoke/supported-os/test.sh:
```diff
@@ -0,0 +1,62 @@
+    healthcheck 1
+    send_events
+    ${ASSERT_EVENTS_FUNC} ${OBSERVER_VERSION}
+    delete_all
```
Should we move this `delete_all` directly after the `terraform_apply`?

IIANM, it is possbile that APM Server is still running (and indexing events/metrics/etc.) here, so we may end up with pollution between test cases. If we do it directly after `terraform_apply`, the previous test case's APM Server should have been stopped.

(Probably warrants a `# comment` if you do make that change.)

kruskall replied:
Oh, good catch! Updated :+1:

axw commented on testing/smoke/supported-os/main.tf:
```diff
@@ -0,0 +1,218 @@
+      "curl ${data.external.getlatestapmserver.result.deb} -o apm-server.deb && curl ${data.external.getlatestapmserver.result.rpm} -o apm-server.rpm",
+      "sudo dpkg -i apm-server.deb || sudo yum -y install apm-server.rpm",
+      "sudo rm /etc/apm-server/apm-server.yml",
+      "echo \"apm-server:\n  host: \\\"0.0.0.0:${local.apm_port}\\\"\noutput:\n  elasticsearch:\n    hosts: [\\\"${module.ec_deployment.elasticsearch_url}\\\"]\n    username: \\\"${module.ec_deployment.elasticsearch_username}\\\"\n    password: \\\"${module.ec_deployment.elasticsearch_password}\\\"\" | sudo tee -a /etc/apm-server/apm-server.yml",
```
Should we be injecting the secret token here too?

kruskall replied:
We don't need to. The APM server is not using the secret token afaik but the output variable is needed because the bash script sends the token in a http header.

axw replied:
OK. I don't think it hurts either way, we can leave it for now.

axw commented on testing/smoke/lib.sh:
```diff
@@ -49,11 +72,12 @@ terraform_destroy() {
         echo "-> Smoke tests FAILED!!"
         echo "-> Printing terraform logs:"
-        cat tf.log
+	#cat tf.log
```
Did you mean to leave this commented out?

kruskall replied:
Yes, I find the previous behaviour really confusing.

Any error during the test assertions is getting suppressed by the tf.log printed after the test failed, so you end up having to scroll to find the error message.

I restored the previous behaviour and we can discuss this in a followup since it's unrelated to the changes.

axw reviewed (approved):
Thanks, this looks much better now. Just a couple of more small issues, then I think this is good to go.

We'll still need to figure out Windows and macOS, but this is obviously much better than what we had ;)  Ideally we would also test aarch64.

axw commented on testing/smoke/supported-os/test.sh:
```diff
@@ -0,0 +1,66 @@
+    append_tfvar "aws_provisioner_key_name" ${KEY_NAME} 1
+    append_tfvar "aws_os" $os
+    terraform_apply
+    #  it is possbile that APM Server is still running (and indexing events/metrics/etc.),
+    #  so we may end up with pollution between test cases. 
+    #  If we do it directly after terraform_apply, the previous test case's APM Server 
+    #  should have been stopped.
```
```suggestion
    # The previous test case's APM Server should have been stopped by now,
    # so there should be no new documents indexed. Delete all existing data.
```

simitt commented:
@kruskall could you fix the linting issues so we can get this merged in?

kruskall commented:
> @kruskall could you fix the linting issues so we can get this merged in?

The linting issue is actually the docker image not being updated: `docker-compose.yml: image docker.elastic.co/elasticsearch/elasticsearch:8.6.0-c9d1a5e9-SNAPSHOT is out of date (latest 'main' snapshot version is 8.7.0-SNAPSHOT)`

It should be fixed by https://github.com/elastic/apm-server/pull/9601
